File: src/core/money.ts
// decimal.js interop (ESM default or namespace)
import * as DecimalModule from 'decimal.js';
const Decimal: any = (DecimalModule as any).default || DecimalModule;

const DEFAULT_SCALE = 2;
const CURRENCY_SCALES: Record<string, number> = {
  USD: 2, EUR: 2, GBP: 2, JPY: 0, CNY: 2, KRW: 0,
};
export const ROUND_HALF_UP = 4; // decimal.js ROUND_HALF_UP

/**
 * Optional scale resolver for non-fiat assets (crypto coins, equity shares, …).
 * Consulted ONLY after the built-in fiat map misses, so the fiat path stays
 * byte-identical. Installed additively via `installAssetScales` in src/instruments.
 *
 * Determinism note (per audit + compaction): this is process-global mutable state.
 * For AI, reproducible, cross-process or "Blueprint First" use: prefer the explicit
 * bypass Money.from(..., scale) or Money.fromWithExplicitScale(value, cur, scale).
 * Never rely on pre-install for determinism; pass + persist the scale with any ledger.
 */
type ScaleResolver = (symbol: string) => number | undefined;
let extraResolver: ScaleResolver | undefined;

/** Register a resolver that supplies decimal scales for non-fiat asset symbols.
 * Prefer explicit scale in Money.from / fromWithExplicitScale for determinism instead of this global. */
export function registerScaleResolver(resolver: ScaleResolver | undefined): void {
  extraResolver = resolver;
}

function scaleFor(currency: string): number {
  const c = currency.toUpperCase();
  if (CURRENCY_SCALES[c] !== undefined) return CURRENCY_SCALES[c]; // fiat wins, unchanged
  const fromResolver = extraResolver?.(c);
  return fromResolver ?? DEFAULT_SCALE;
}

/**
 * Central guard for every external numeric entry point. decimal.js accepts the
 * strings "Infinity"/"-Infinity"/"NaN" (and the number forms), which silently poison
 * all downstream arithmetic and comparisons (NaN comparisons are always false, so
 * positive/scale/balance guards pass). Reject them at the door so a non-finite value
 * can never enter a Money amount, an FX rate, a scalar, or an allocation ratio.
 */
function toFiniteDecimal(value: string | number, context: string): any {
  if (typeof value === 'number' && !Number.isFinite(value)) {
    throw new Error(`${context}: non-finite number ${value}`);
  }
  const dec = new Decimal(String(value));
  if (!dec.isFinite()) {
    throw new Error(`${context}: must be a finite number; got "${value}"`);
  }
  return dec;
}

/**
 * Guard the RESULT of an operation. decimal.js division by zero yields Infinity
 * (it does not throw), and an Infinity/NaN result would silently poison every
 * downstream comparison (all false for non-finite). Reject it so a Money can
 * never hold a non-finite amount, even though its inputs were finite.
 */
function assertFiniteResult(dec: any, context: string): any {
  if (!dec.isFinite()) throw new Error(`${context}: non-finite result`);
  return dec;
}

/**
 * FXRate(from, to, rate, asOf?, source?). Rate is an exact conversion factor (not a monetary amount).
 * Stored exactly as a decimal string — never a float, never parseFloat — so high-precision rates
 * survive intact. Used only for explicit multi-currency legs.
 */
export class FXRate {
  public readonly from: string;
  public readonly to: string;
  public readonly rate: string; // exact canonical decimal string (no float)
  private readonly _rate: any;  // Decimal instance
  public readonly asOf?: string;
  public readonly source?: string;

  constructor(from: string, to: string, rate: number | string, asOf?: string, source?: string) {
    this.from = from.toUpperCase();
    this.to = to.toUpperCase();
    // Reject Infinity/NaN from BOTH number and string forms (string "Infinity" otherwise
    // slips past a number-only check and yields an infinite rate that poisons convert()).
    this._rate = toFiniteDecimal(rate, 'FXRate'); // exact; String() captures full digits, no parseFloat
    this.rate = this._rate.toString();
    this.asOf = asOf;
    this.source = source;
  }

  /** Exact Decimal form of the rate, for exact multiplication. */
  rateDecimal(): any { return this._rate; }

  toString(): string {
    return `1 ${this.from} = ${this.rate} ${this.to}${this.asOf ? ' @ ' + this.asOf : ''}`;
  }
}

export class Money {
  private readonly _amount: any; // Decimal instance - exact arbitrary precision
  public readonly currency: string;
  public readonly scale: number;
  public readonly asOf?: string;
  public readonly provenance?: string;

  private constructor(amount: any, currency: string, scale?: number, asOf?: string, provenance?: string) {
    this._amount = amount;
    this.currency = currency.toUpperCase();
    this.scale = scale ?? scaleFor(this.currency);
    this.asOf = asOf;
    this.provenance = provenance;
  }

  /**
   * Money.from(value, currency). String coercion prevents float traps. Optional scale override
   * (5th param) bypasses the global resolver entirely for determinism (see fromWithExplicitScale).
   * For AI/financial reproducibility, always supply explicit scale for non-fiat when possible.
   */
  static from(value: string | number, currency: string, asOf?: string, provenance?: string, scale?: number): Money {
    // VULN-02: isSafeInteger rejects both non-integers (0.1) and values above MAX_SAFE_INTEGER
    // where String() already returns the wrong number.
    if (typeof value === 'number' && !Number.isSafeInteger(value)) {
      throw new Error(`Money.from: number ${value} is non-integer or outside safe-integer range — pass a string to preserve precision`);
    }
    // VULN-01: Decimal supports "Infinity" / "NaN" as strings; reject non-finite input before
    // any guard that relies on arithmetic comparisons (all false for NaN/Infinity).
    const dec = toFiniteDecimal(value, 'Money.from');
    const resolvedScale = scale ?? scaleFor(currency.toUpperCase());
    if (dec.decimalPlaces() > resolvedScale) {
      throw new Error(
        `Money.from: ${value} ${currency.toUpperCase()} has ${dec.decimalPlaces()} decimal places but the currency scale is ${resolvedScale} — use a string with at most ${resolvedScale} decimal places`
      );
    }
    return new Money(dec, currency, scale, asOf, provenance);
  }

  /**
   * Explicit-scale factory that *never* consults the global/process resolver.
   * Use for determinism in AI agents, cross-process, or when you want to avoid installAssetScales side-effects.
   * The scale is mandatory and used verbatim for validation + storage.
   */
  static fromWithExplicitScale(value: string | number, currency: string, scale: number, asOf?: string, provenance?: string): Money {
    if (typeof scale !== 'number' || !Number.isFinite(scale) || scale < 0) {
      throw new Error('fromWithExplicitScale: scale must be a finite non-negative number');
    }
    return Money.from(value, currency, asOf, provenance, scale);
  }

  /** Zero for currency (preferred over from(0)). */
  static zero(currency: string, asOf?: string, provenance?: string): Money {
    return Money.from(0, currency, asOf, provenance);
  }

  equals(other: Money): boolean {
    return this.currency === other.currency && this._amount.eq(other._amount);
  }

  isZero(): boolean { return this._amount.isZero(); }

  add(other: Money): Money {
    if (this.currency !== other.currency) {
      throw new Error(`Currency mismatch: ${this.currency} vs ${other.currency}`);
    }
    return new Money(
      this._amount.add(other._amount),
      this.currency,
      this.scale,
      this.asOf ?? other.asOf,
      Money.combineProvenance(this.provenance, other.provenance)
    );
  }

  sub(other: Money): Money {
    if (this.currency !== other.currency) {
      throw new Error(`Currency mismatch: ${this.currency} vs ${other.currency}`);
    }
    return new Money(
      this._amount.sub(other._amount),
      this.currency,
      this.scale,
      this.asOf ?? other.asOf,
      Money.combineProvenance(this.provenance, other.provenance)
    );
  }

  /** mul(scalar) — roundingMode passed to toDecimalPlaces if provided. */
  mul(scalar: string | number, roundingMode?: number): Money {
    let r = assertFiniteResult(this._amount.mul(toFiniteDecimal(scalar, 'Money.mul')), 'Money.mul');
    if (roundingMode !== undefined) r = r.toDecimalPlaces(this.scale, roundingMode);
    return new Money(r, this.currency, this.scale, this.asOf, this.provenance);
  }

  /** div(scalar) — optional roundingMode. Division by zero is rejected (no Infinity). */
  div(scalar: string | number, roundingMode?: number): Money {
    const d = toFiniteDecimal(scalar, 'Money.div');
    if (d.isZero()) throw new Error('Money.div: division by zero');
    let r = assertFiniteResult(this._amount.div(d), 'Money.div');
    if (roundingMode !== undefined) r = r.toDecimalPlaces(this.scale, roundingMode);
    return new Money(r, this.currency, this.scale, this.asOf, this.provenance);
  }

  /** Convert via explicit FXRate. Exact multiply, then round to the target currency scale. */
  convert(rate: FXRate, roundingMode: number = ROUND_HALF_UP): Money {
    if (this.currency !== rate.from) throw new Error(`FX mismatch: ${this.currency} vs ${rate.from}`);
    const targetScale = scaleFor(rate.to);
    const amt = assertFiniteResult(
      this._amount.times(rate.rateDecimal()).toDecimalPlaces(targetScale, roundingMode),
      'Money.convert',
    );
    return new Money(amt, rate.to, targetScale, rate.asOf, rate.source || this.provenance);
  }

  /** toFormat({decimals?, symbol?}). */
  toFormat(opts: { decimals?: number; symbol?: string } = {}): string {
    const d = opts.decimals ?? this.scale;
    const s = opts.symbol ? opts.symbol + ' ' : '';
    return `${s}${this._amount.toFixed(d)} ${this.currency}`;
  }

  /**
   * allocate(ratios) — exact split, remainder to last. Supports string|number ratios.
   * Always sums to original (kernel invariant).
   */
  allocate(ratios: (string | number)[]): Money[] {
    if (ratios.length === 0) return [];
    const ps = ratios.map(r => toFiniteDecimal(r, 'Money.allocate'));
    if (ps.some(p => p.isNegative())) {
      throw new Error('Money.allocate: ratios must be non-negative');
    }
    const tot = ps.reduce((a, b) => a.add(b), new Decimal(0));
    if (tot.isZero()) {
      throw new Error('Money.allocate: ratios sum to zero');
    }

    const res: Money[] = [];
    let alloc = new Decimal(0);
    for (let i = 0; i < ps.length; i++) {
      const share = (i === ps.length - 1)
        ? this._amount.sub(alloc)
        : this._amount.mul(ps[i]).div(tot).toDecimalPlaces(this.scale, 1);
      const m = new Money(share, this.currency, this.scale, this.asOf, this.provenance);
      res.push(m);
      alloc = alloc.add(share);
    }
    return res;
  }

  toString(): string {
    return `${this._amount.toFixed(this.scale)} ${this.currency}`;
  }

  toDecimal(): any { return this._amount; }

  /** Hash for determinism verification. */
  toHashable(): string {
    return `${this._amount.toString()}:${this.currency}:${this.scale}:${this.asOf ?? ''}`;
  }

  /** Compare to other (same currency required). Returns -1, 0, or 1. */
  compare(other: Money): -1 | 0 | 1 {
    if (this.currency !== other.currency) {
      throw new Error(`Currency mismatch: ${this.currency} vs ${other.currency}`);
    }
    return this._amount.cmp(other._amount);
  }

  /** Negate the amount (sign flip), preserve currency, scale, asOf, provenance. */
  negate(): Money {
    return new Money(this._amount.negated(), this.currency, this.scale, this.asOf, this.provenance);
  }

  /** Absolute value, preserve currency, scale, asOf, provenance. */
  abs(): Money {
    return new Money(this._amount.abs(), this.currency, this.scale, this.asOf, this.provenance);
  }

  /**
   * Serialize to plain object for roundtrips, persistence, hashing.
   * v: version for forward compat.
   */
  toJSON(): { v: string; a: string; c: string; s: number; asOf?: string; provenance?: string } {
    return {
      v: '1',
      a: this._amount.toString(),
      c: this.currency,
      s: this.scale, // persist scale so asset Money rehydrates correctly without the global resolver
      asOf: this.asOf,
      provenance: this.provenance,
    };
  }

  /**
   * Reconstruct Money from the toJSON shape.
   * Accepts loose keys for flexibility (a/amount, c/currency, provenance/p).
   */
  static fromJSON(j: any): Money {
    if (!j || typeof j !== 'object') throw new Error('Money.fromJSON expects object');
    const amt = j.a ?? j.amount;
    const cur = j.c ?? j.currency;
    const prov = j.provenance ?? j.p;
    if (amt == null || !cur) {
      throw new Error('Money.fromJSON missing amount or currency');
    }
    const scale = typeof j.s === 'number' ? j.s : undefined; // backward compatible: v1 JSON without `s`
    return Money.from(amt, cur, j.asOf, prov, scale);
  }

  /** Combine provenance strings for add/sub (internal, exact, no mutation). */
  private static combineProvenance(a?: string, b?: string): string | undefined {
    if (a && b && a !== b) return `${a}|${b}`;
    return a ?? b;
  }
}

File: reference-implementations/python/ledger/money.py
"""
Python canonical reference implementation of ledger Money + FXRate.

Faithful port of TS src/core/money.ts semantics:
- Exact arithmetic via decimal.Decimal (no float for monetary amounts).
- Money.from: non-integer float is rejected with explicit error.
- Scales per currency for minor-unit validation (not for internal storage).
- Provenance tracking, asOf.
- Immutable.
- Full roundtrip JSON.
- Supports mul/div scalar, allocate (exact + remainder rule), convert via FXRate.
"""

from decimal import Decimal, ROUND_HALF_UP, InvalidOperation, getcontext
from typing import Optional, Any, Union, List
import json

def _quantize(d: Decimal, scale: int, rounding: Any = None) -> Decimal:
    """Helper to quantize to currency scale, mirroring decimal.js toDecimalPlaces behavior."""
    if rounding is None:
        rounding = ROUND_HALF_UP
    q = Decimal('1.' + '0' * scale) if scale > 0 else Decimal('1')
    return d.quantize(q, rounding=rounding)

# Mirror TS: keep high precision for internal; scale is for formatting + sub-scale guard in journal.
getcontext().prec = 50  # ample for financial

DEFAULT_SCALE = 2
CURRENCY_SCALES: dict[str, int] = {
    "USD": 2, "EUR": 2, "GBP": 2, "JPY": 0, "CNY": 2, "KRW": 0,
    "USDC": 2, "USDT": 2, "BTC": 8, "ETH": 8,
}

_extra_resolver = None  # type: Optional[callable]


def register_scale_resolver(resolver: Optional[callable]) -> None:
    global _extra_resolver
    _extra_resolver = resolver


def scale_for(currency: str) -> int:
    c = currency.upper()
    if c in CURRENCY_SCALES:
        return CURRENCY_SCALES[c]
    if _extra_resolver is not None:
        v = _extra_resolver(c)
        if v is not None:
            return v
    return DEFAULT_SCALE


ROUNDING = ROUND_HALF_UP


class FXRate:
    """Exact rate (not monetary). Stored as decimal string. Mirror of TS FXRate."""

    def __init__(self, from_: str, to: str, rate: Union[str, int, float, Decimal], as_of: Optional[str] = None, source: Optional[str] = None):
        self.from_ = from_.upper()
        self.to = to.upper()
        if isinstance(rate, float) and not rate.is_integer():
            # Still accept for rates (they are factors), but document; TS uses String(rate) always.
            pass
        try:
            self._rate = Decimal(str(rate))
        except (InvalidOperation, ValueError) as e:
            raise ValueError(f"Invalid FX rate {rate}") from e
        self.rate = str(self._rate)
        self.as_of = as_of
        self.source = source

    def rate_decimal(self) -> Decimal:
        return self._rate

    def __str__(self) -> str:
        s = f"1 {self.from_} = {self.rate} {self.to}"
        if self.as_of:
            s += f" @ {self.as_of}"
        return s


class Money:
    def __init__(self, amount: Decimal, currency: str, scale: Optional[int] = None,
                 as_of: Optional[str] = None, provenance: Optional[str] = None):
        if not isinstance(amount, Decimal):
            amount = Decimal(str(amount))
        self._amount: Decimal = amount
        self.currency: str = currency.upper()
        self.scale: int = scale if scale is not None else scale_for(self.currency)
        self.as_of = as_of
        self.provenance = provenance

    @classmethod
    def from_(cls, value: Any, currency: str, as_of: Optional[str] = None,
              provenance: Optional[str] = None, scale: Optional[int] = None) -> "Money":
        """
        Strict construction. Mirrors TS:
        - Reject raw non-integer float (IEEE754 trap).
        - Coerce via str() to Decimal (captures full digits).
        """
        if isinstance(value, float) and not value.is_integer():
            raise ValueError(
                f"Money.from_ forbids raw float {value} — IEEE754 error. "
                f"Pass str or int or Decimal, e.g. '{value}'"
            )
        try:
            if isinstance(value, Decimal):
                dec = value
            else:
                dec = Decimal(str(value))
            # Do not auto-quantize here (preserve precision like TS); journal validate guards sub-scale.
            return cls(dec, currency, scale, as_of, provenance)
        except (InvalidOperation, ValueError) as e:
            raise ValueError(f"Invalid monetary value {value}") from e

    @classmethod
    def zero(cls, currency: str, as_of: Optional[str] = None, provenance: Optional[str] = None) -> "Money":
        return cls.from_(0, currency, as_of, provenance)

    def to_decimal(self) -> Decimal:
        return self._amount

    def equals(self, other: "Money") -> bool:
        return self.currency == other.currency and self._amount == other._amount

    def is_zero(self) -> bool:
        return self._amount.is_zero()

    def add(self, other: "Money") -> "Money":
        if self.currency != other.currency:
            raise ValueError(f"Currency mismatch: {self.currency} vs {other.currency}")
        new_prov = self._combine_provenance(self.provenance, other.provenance)
        return Money(self._amount + other._amount, self.currency, self.scale, self.as_of or other.as_of, new_prov)

    def sub(self, other: "Money") -> "Money":
        if self.currency != other.currency:
            raise ValueError(f"Currency mismatch: {self.currency} vs {other.currency}")
        new_prov = self._combine_provenance(self.provenance, other.provenance)
        return Money(self._amount - other._amount, self.currency, self.scale, self.as_of or other.as_of, new_prov)

    def mul(self, scalar: Union[str, int, float, Decimal], rounding_mode: Any = None) -> "Money":
        r = self._amount * Decimal(str(scalar))
        if rounding_mode is not None:
            r = _quantize(r, self.scale, rounding_mode)
        return Money(r, self.currency, self.scale, self.as_of, self.provenance)

    def div(self, scalar: Union[str, int, float, Decimal], rounding_mode: Any = None) -> "Money":
        r = self._amount / Decimal(str(scalar))
        if rounding_mode is not None:
            r = _quantize(r, self.scale, rounding_mode)
        return Money(r, self.currency, self.scale, self.as_of, self.provenance)

    def convert(self, rate: FXRate, rounding_mode: Any = ROUND_HALF_UP) -> "Money":
        if self.currency != rate.from_:
            raise ValueError(f"FX mismatch: {self.currency} vs {rate.from_}")
        target_scale = scale_for(rate.to)
        amt = _quantize(self._amount * rate.rate_decimal(), target_scale, rounding_mode)
        return Money(amt, rate.to, target_scale, rate.as_of, rate.source or self.provenance)

    def allocate(self, ratios: List[Union[str, int, float, Decimal]]) -> List["Money"]:
        if not ratios:
            return []
        ps = [Decimal(str(r)) for r in ratios]
        total = sum(ps, Decimal(0))
        if total.is_zero():
            return [Money.zero(self.currency, self.as_of, self.provenance) for _ in ratios]
        res: List[Money] = []
        alloc = Decimal(0)
        for i, p in enumerate(ps):
            if i == len(ps) - 1:
                share = self._amount - alloc
            else:
                share = _quantize((self._amount * p / total), self.scale)
            m = Money(share, self.currency, self.scale, self.as_of, self.provenance)
            res.append(m)
            alloc += share
        return res

    def negate(self) -> "Money":
        return Money(self._amount.copy_negate(), self.currency, self.scale, self.as_of, self.provenance)

    def abs(self) -> "Money":
        return Money(self._amount.copy_abs(), self.currency, self.scale, self.as_of, self.provenance)

    def compare(self, other: "Money") -> int:
        if self.currency != other.currency:
            raise ValueError(f"Currency mismatch: {self.currency} vs {other.currency}")
        return (self._amount > other._amount) - (self._amount < other._amount)

    def to_format(self, decimals: Optional[int] = None, symbol: Optional[str] = None) -> str:
        d = decimals if decimals is not None else self.scale
        s = f"{symbol} " if symbol else ""
        q = _quantize(self._amount, d)
        return f"{s}{q} {self.currency}"

    def __str__(self) -> str:
        # Mirror primary toString using scale
        try:
            q = _quantize(self._amount, self.scale)
            return f"{q} {self.currency}"
        except Exception:
            return f"{self._amount} {self.currency}"

    def to_hashable(self) -> str:
        return f"{str(self._amount)}:{self.currency}:{self.scale}:{self.as_of or ''}"

    def to_json(self) -> dict:
        return {
            "v": "1",
            "a": str(self._amount),
            "c": self.currency,
            "s": self.scale,  # persist scale so asset Money rehydrates correctly without the global resolver
            "asOf": self.as_of,
            "provenance": self.provenance,
        }

    @classmethod
    def from_json(cls, j: dict) -> "Money":
        if not isinstance(j, dict):
            raise ValueError("Money.from_json expects dict")
        amt = j.get("a") or j.get("amount")
        cur = j.get("c") or j.get("currency")
        prov = j.get("provenance") or j.get("p")
        if amt is None or not cur:
            raise ValueError("Money.from_json missing amount or currency")
        s = j.get("s")
        scale = s if isinstance(s, int) and not isinstance(s, bool) else None  # backward compatible: legacy JSON without `s`
        return cls.from_(amt, cur, j.get("asOf"), prov, scale)

    @staticmethod
    def _combine_provenance(a: Optional[str], b: Optional[str]) -> Optional[str]:
        if a and b and a != b:
            return f"{a}|{b}"
        return a or b

    def __repr__(self) -> str:
        return f"Money({self})"


def money_from(value: Any, currency: str, **kw) -> Money:
    return Money.from_(value, currency, **kw)
